File: generated_code/cpp/solution_1093_2.cpp
#include <string>
#include <vector>
#include <cctype>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* recoverFromPreorder(string traversal) {
        // A stack to keep track of the current path from the root down to the last processed node.
        // The size of the stack at any point matches the depth of the node being evaluated.
        vector<TreeNode*> path;
        int i = 0;
        int n = traversal.length();
        
        while (i < n) {
            // Count the number of dashes to determine the depth of the current node
            int depth = 0;
            while (i < n && traversal[i] == '-') {
                depth++;
                i++;
            }
            
            // Extract the value of the current node
            int val = 0;
            while (i < n && isdigit(traversal[i])) {
                val = val * 10 + (traversal[i] - '0');
                i++;
            }
            
            // Create the new node
            TreeNode* node = new TreeNode(val);
            
            // If the stack size is greater than the current depth, it means we have finished 
            // processing the subtrees of the deeper nodes, so we pop them until we reach the parent.
            while (path.size() > depth) {
                path.pop_back();
            }
            
            // If the path is not empty, attach the current node to its parent
            if (!path.empty()) {
                if (path.back()->left == nullptr) {
                    // The problem guarantees that if a node has only one child, it's the left child.
                    path.back()->left = node;
                } else {
                    path.back()->right = node;
                }
            }
            
            // Push the current node onto the stack path
            path.push_back(node);
        }
        
        // The first node pushed to the stack is the root of the tree
        return path.empty() ? nullptr : path.front();
    }
};

File: generated_code/python/solution_1093_2.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def recoverFromPreorder(self, traversal: str) -> Optional[TreeNode]:
        """
        Algorithm: Iterative Preorder Simulation using a Stack
        Time Complexity: O(N) where N is the length of the traversal string. 
                         Each character is processed a constant number of times.
        Space Complexity: O(H) where H is the maximum depth of the tree (stack size).
                          In the worst case, H can be equal to the number of nodes.
        """
        stack: List[TreeNode] = []
        i = 0
        n = len(traversal)
        
        while i < n:
            # 1. Count the number of dashes to determine the depth of the current node
            depth = 0
            while i < n and traversal[i] == '-':
                depth += 1
                i += 1
            
            # 2. Extract the integer value of the current node
            val = 0
            while i < n and traversal[i].isdigit():
                val = val * 10 + int(traversal[i])
                i += 1
                
            node = TreeNode(val)
            
            # 3. Maintain the stack so its size matches the required depth of the current node.
            # If the stack size is greater than the depth, it means we have finished processing 
            # the current subtree, so we pop elements until we reach the parent level.
            while len(stack) > depth:
                stack.pop()
                
            # 4. Attach the node to its parent.
            # Preorder traversal ensures the left child is always populated before the right child.
            if stack:
                if not stack[-1].left:
                    stack[-1].left = node
                else:
                    stack[-1].right = node
                    
            # 5. Push the current node onto the stack as a potential parent for subsequent nodes
            stack.append(node)
            
        # The root of the reconstructed tree will always be the first element in the stack
        return stack[0] if stack else None